Create rocketChat djangoapp

andrey-canon opened the pull request:
## Description 
Create a new django app in order to replace discussion, summarizing the changes:
- Create a new view with rocketchat operations (Create and log a user)
- Add new template
- Add new urls 
- Make discussions microsite aware 
- Add a new rocketchat tab to the course
## Reviewers 
- [x] @jfavellar90 
- [ ] @diegomillan 
- [ ] @nicovanniekerk

jfavellar90 commented on common/djangoapps/rocket_chat/views.py:
```diff
@@ -0,0 +1,111 @@
+from rocketchat_API.rocketchat import RocketChat as ApiRocketChat
+
+
+def rocket_chat_discussion(request, course_id):
```
login_required decorator for the view

andrey-canon replied:
check

jfavellar90 commented on requirements/edx/github.txt:
```diff
@@ -114,4 +114,5 @@ git+https://github.com/proversity-org/agnostic-content-xblock.git#egg=agnostic-c
 
-git+https://github.com/edx/edx-analytics-data-api-client.git#egg=edx-analytics-data-api-client
\ No newline at end of file
+git+https://github.com/edx/edx-analytics-data-api-client.git#egg=edx-analytics-data-api-client
```
what is this requirement for? do you need analytics data api client?

andrey-canon replied:
that is an old line, I just added line  118

jfavellar90 replied:
OK

jfavellar90 commented on lms/djangoapps/discussion/views.py:
```diff
@@ -237,6 +238,10 @@ def forum_form_discussion(request, course_key):
     Renders the main Discussion page, potentially filtered by a search query
     """
+
+    if not configuration_helpers.get_value('ENABLE_DISCUSSION_SERVICE', settings.FEATURES.get('ENABLE_DISCUSSION_SERVICE')):
```
Should we use the is_discussion_enabled method you modified in the file below instead of calling again the configuration helpers?

andrey-canon replied:
I will check this

jfavellar90 commented on lms/djangoapps/django_comment_client/utils.py:
```diff
@@ -1020,7 +1021,7 @@ def is_discussion_enabled(course_id):
     Return True if discussions are enabled; else False
     """
-    return settings.FEATURES.get('ENABLE_DISCUSSION_SERVICE')
+    return configuration_helpers.get_value('ENABLE_DISCUSSION_SERVICE', settings.FEATURES.get('ENABLE_DISCUSSION_SERVICE'))
```
Question, what is the normal behavior when discussions are disabled? discussion tab is not displayed on courses? does it have any real effect on the platform (with the default edx-platform behavior)?

andrey-canon replied:
The tab is not displayed and the discussions' urls are not loaded

jfavellar90 commented on common/djangoapps/rocket_chat/views.py:
```diff
@@ -0,0 +1,111 @@
+from xmodule.modulestore.django import modulestore
+from openedx.core.djangoapps.site_configuration import helpers as configuration_helpers
+
+from rocketchat_API.rocketchat import RocketChat as ApiRocketChat
```
Here on the platform we use the python rocketchat API but in the xblock we're using our own API, this is maybe OK for now but we need to keep in mind that is better if we use the same API everywhere

andrey-canon replied:
Yes, you are right, this was also a test of compatibility, because the API is for python 3 and for the xblock there are methods the API does not have, I think the better is extend this API on the xblock in order to reduce code and centralize an API.

jfavellar90 replied:
The test of compatibility you mention is checking if this python 3 package does not fail under our case of use? could it be a bit dangerous in the case we find a failure?

jfavellar90 commented on common/djangoapps/rocket_chat/views.py:
```diff
@@ -0,0 +1,111 @@
+        staff_access = has_access(user, 'staff', course)
+        user_is_enrolled = CourseEnrollment.is_enrolled(user, course.id)
+
+        course_homepage_invert_title =\
```
backslash???

andrey-canon replied:
explicit line joining, it is not necessary

andrey-canon commented:
@jfavellar90  please check

jfavellar90 commented:
My concern here is the fact of having a dependency on python 3. This can fail later using other API methods @andrey-canon

andrey-canon commented:
@jfavellar90 Could you check the last commit ?

jfavellar90 commented on common/djangoapps/rocket_chat/utils.py:
```diff
@@ -0,0 +1,47 @@
+            api_rocket_chat.groups_create(name=room_name, **kwargs)
+
+    except AttributeError:
+        pass
```
Should we log something here saying that the group creation failed?

andrey-canon replied:
ok

jfavellar90 commented on common/djangoapps/rocket_chat/views.py:
```diff
@@ -0,0 +1,93 @@
+            'user_is_enrolled': user_is_enrolled,
+        }
+
+        rocket_chat_settings = get_rocket_chat_settings()
```
What if this method returns none (because the configuration is not correct)? or even if the keys "admin_user", "admin_pass" or "public_url_service" are not defined? should we render a template saying there was an error and log something? or raise an ImproperlyConfigured exception (https://docs.djangoproject.com/en/1.8/ref/exceptions/#improperlyconfigured)?

jfavellar90 replied:
What do you prefer, rendering an alternative html saying that rocket chat is not available and also logging something? or raising the exception?

andrey-canon replied:
Currently if the settings are None the view will show a message (line 91)  I am going to add somo logs to track the possible problems

jfavellar90 replied:
Yap you're right, what an idiot, the not covered case is when the settings are not well defined (some of the required keys is not present)

andrey-canon replied:
ok

jfavellar90 commented on common/djangoapps/rocket_chat/views.py:
```diff
@@ -0,0 +1,93 @@
+
+            return render_to_response('rocket_chat/rocket_chat.html', context)
+
+        context['rocket_chat_error_message'] = 'RocketChat settings does not exist'
```
Looking to this message, I prefer to send a different one in the context like "Rocket chat service is currently not available" or something, and use the message "RocketChat settings does not exist" in a log, because I think this message is rather for us the devs, the users don't care about the fact that rocket chat settings exist or not

jfavellar90 replied:
@andrey-canon

andrey-canon replied:
ok

andrey-canon committed:
Add utils file with the following methods:
- create course group
- create user
- get rocketchat settings
Addressing PR comments
Add new rocketchat tab and make discussions micrositeaware
Create a new django app in order to replace discussions

andrey-canon commented:
@jfavellar90 please check the last commit
